### PR TITLE
KAFKA-16629: Add broker-related tests to ConfigCommandIntegrationTest

### PR DESCRIPTION
**Pull request by m1a2st**
wip, add new KRAFT test for ConfigCommandIntegrationTest

### CODE REVIEW

**Comment by m1a2st** [comment]
Because Old Test Only have tests with Zookeeper, thus I change old test for Zookeeper and Kraft, and add new tests for KRaft, I mainly test the the config update mode for configs

**Comment by m1a2st** [comment]
Hello @chia7712 please review, Thank you

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 1]
```diff
@@ -172,72 +138,210 @@ public void testDynamicBrokerConfigUpdateUsingZooKeeper() throws Exception {
...
+        for (Map<String, String> config : configs) {
+            assertEquals(config, entityConfigs);
+        }
+    }
+
+    private void alterConfigWithZk(KafkaZkClient zkClient, Optional<String> brokerId, Map<String, String>... configs) {
+        String configStr = transferConfigMapsToString(configs);
+        ConfigCommand.ConfigCommandOptions addOpts =
+                new ConfigCommand.ConfigCommandOptions(toArray(alterOpts, entityOp(brokerId), asList("--add-config", configStr)));
+        ConfigCommand.alterConfigWithZk(zkClient, addOpts, adminZkClient);
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:439]: Could we convert `adminZkClient` to a local variable here? A class-level `adminZkClient` is error-prone in virtue of NPE

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 2]
```diff
@@ -172,72 +138,210 @@ public void testDynamicBrokerConfigUpdateUsingZooKeeper() throws Exception {
...
+        verifyConfig(zkClient, brokerId, configs);
+    }
+
+    @ClusterTests({
+            @ClusterTest(clusterType = Type.KRAFT),
+            @ClusterTest(clusterType = Type.CO_KRAFT)
+    })
+    public void testDynamicBrokerConfigUpdateUsingKraft() {
+        alterOpts = generateKraftAlterOpts();
+        Admin client = cluster.createAdminClient();
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: Could you use `try-resource` to make sure `admin` get closed?

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 3]
```diff
@@ -248,26 +352,75 @@ private void registerBrokerInZk(KafkaZkClient zkClient, int id) {
...
+                new ConfigCommand.ConfigCommandOptions(toArray(alterOpts, entityOp(brokerId),
+                        asList("--delete-config", String.join(",", singleton))));
+        ConfigCommand.alterConfig(client, deleteOpts);
+        verifyConfig(client, brokerId, new Map[]{Collections.emptyMap()});
+    }
+
     @SafeVarargs
-    static <T> Seq<T> seq(T...seq) {
-        return seq(Arrays.asList(seq));
+    static <T> Seq<T> seq(T... seq) {
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: It seems we need this function to pass single `endpoint` to create  `Broker`. Maybe we can add a new constructor to `Broker` to accept single `endpoint`. WDYT?

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 4]
```diff
@@ -283,8 +436,26 @@ private static String captureStandardStream(boolean isErr, Runnable runnable) {
                 System.setErr(currentStream);
             else
                 System.setOut(currentStream);
+        }
+    }
+
+    private static class BrokerCommand implements Runnable {
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: This can be simplified to a lambda function. For example:
```java
    private static Runnable run(Stream<String> command) {
        return () -> {
            try {
                ConfigCommand.main(command.toArray(String[]::new));
            } catch (RuntimeException e) {
                // do nothing.
            } finally {
                Exit.resetExitProcedure();
            }
        };
    }
```

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 5]
```diff
@@ -172,72 +138,210 @@ public void testDynamicBrokerConfigUpdateUsingZooKeeper() throws Exception {
...
-        assertThrows(IllegalArgumentException.class, () -> alterConfigWithZk(zkClient, Collections.singletonMap("message.max.size", "220000"), Optional.empty()));
+        assertThrows(IllegalArgumentException.class, () -> alterConfigWithZk(zkClient, Optional.of(brokerId), singletonMap("message.max.size", "210000")));
+        assertThrows(IllegalArgumentException.class, () -> alterConfigWithZk(zkClient, Optional.empty(), singletonMap("message.max.size", "220000")));
 
         // Dynamic config updates using ZK should for a different broker that is not running should succeed
-        alterAndVerifyConfig(zkClient, Collections.singletonMap("message.max.size", "230000"), Optional.of("2"));
+        alterAndVerifyConfig(zkClient, Optional.of("2"), singletonMap("message.max.size", "230000"));
+    }
+
+    private void alterAndVerifyConfig(KafkaZkClient zkClient, Optional<String> brokerId, Map<String, String>... configs) {
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: We should avoid using variages for collection types. The callers should merge those collections into single one.

**Review by chia7712 — COMMENTED** [review]
@m1a2st Thanks for this great patch!

**Comment by chia7712** [comment]
@m1a2st please fix the build error

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 6]
```diff
@@ -171,120 +144,352 @@ public void testDynamicBrokerConfigUpdateUsingZooKeeper() throws Exception {
...
+
+        String errOut = captureStandardMsg(run(args));
+
+        checkErrOut.accept(errOut);
+        assertNotNull(exitStatus.get());
+        assertEquals(1, exitStatus.get());
+    }
+
+    private Stream<String> quorumArgs() {
+        return cluster.isKRaftTest()
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:417]: We need to test `--bootstrap-server` Even though the cluster is zk mode

**Comment by m1a2st** [comment]
@chia7712 , I have been add zookeeper `--bootstrap-server` tests, please review, thank you

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 7]
```diff
@@ -63,13 +73,26 @@
 
 @SuppressWarnings("deprecation") // Added for Scala 2.12 compatibility for usages of JavaConverters
 @ExtendWith(value = ClusterTestExtensions.class)
+@ClusterTestDefaults
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: If you don't want to change any default value, this can get removed

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 8]
```diff
@@ -63,13 +73,26 @@
...
 
+    private List<String> alterOpts;
+    private final String defaultBrokerId = "0";
     private final ClusterInstance cluster;
 
+    private static Runnable run(Stream<String> command) {
+        return () -> {
+            try {
+                ConfigCommand.main(command.toArray(String[]::new));
+            } catch (RuntimeException e) {
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:85]: This is redundant if you do nothing in catching.

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 9]
```diff
@@ -82,202 +105,448 @@ public void testExitWithNonZeroStatusOnUpdatingUnallowedConfig() {
             "--alter",
             "--add-config", "security.inter.broker.protocol=PLAINTEXT")),
             errOut ->
-                assertTrue(errOut.contains("Cannot update these configs dynamically: Set(security.inter.broker.protocol)"), errOut));
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:85]: Please revert this unrelated change. BTW, this test case should run all cluster types, right?

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for this nice patch. Please take a look at following comments.

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java (hunk 8 above)
Comment (m1a2st) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:85]: If removed this catch block, the `testExitWithNonZeroStatusOnUpdatingUnallowedConfig` can't test the message, because the RuntimeException doesn't been catched

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java (hunk 9 above)
Comment (m1a2st) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:85]: True, this test can run at all type cluster tests

**Comment by chia7712** [comment]
```
错误
org.apache.kafka.common.errors.InvalidConfigurationException: All sensitive broker config entries must be specified for --alter, missing entries: Set(message.max.size)
栈跟踪
org.apache.kafka.common.errors.InvalidConfigurationException: All sensitive broker config entries must be specified for --alter, missing entries: Set(message.max.size)
```
@m1a2st Could you please take a look at above failures?

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 10]
```diff
@@ -65,219 +74,475 @@
...
     }
 
-    @SuppressWarnings({"deprecation"})
-    static <T> Seq<T> seq(Collection<T> seq) {
-        return JavaConverters.asScalaIteratorConverter(seq.iterator()).asScala().toSeq();
+    private void verifyConfig(Admin client, Optional<String> brokerId, Map<String, String> config) throws RuntimeException {
+        ConfigResource configResource = new ConfigResource(ConfigResource.Type.BROKER, brokerId.orElse(defaultBrokerId));
+        client.describeConfigs(singletonList(configResource))
+                .all()
+                .whenComplete((c, e) -> {
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: This will swallow the exception since it is running in `whenComplete`

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 11]
```diff
@@ -65,219 +74,479 @@
...
         String brokerId = "1";
-        adminZkClient = new AdminZkClient(zkClient, scala.None$.empty());
-        alterOpts = Arrays.asList("--zookeeper", zkConnect, "--entity-type", "brokers", "--alter");
+        AdminZkClient adminZkClient = new AdminZkClient(zkClient, scala.None$.empty());
+        alterOpts = asList("--zookeeper", zkConnect, "--entity-type", "brokers", "--alter");
 
         // Add config
-        alterAndVerifyConfig(zkClient, Collections.singletonMap("message.max.size", "110000"), Optional.of(brokerId));
-        alterAndVerifyConfig(zkClient, Collections.singletonMap("message.max.size", "120000"), Optional.empty());
+        alterAndVerifyConfig(zkClient, adminZkClient, Optional.of(brokerId),
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:138]: `message.max.size` is not a correct config in kafka. Maybe we should `message.max.bytes` instead

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 12]
```diff
@@ -65,219 +74,479 @@
...
+        String configStr = transferConfigMapToString(config);
+        ConfigCommand.ConfigCommandOptions addOpts =
+                new ConfigCommand.ConfigCommandOptions(toArray(alterOpts, entityOp(brokerId), asList("--add-config", configStr)));
+        ConfigCommand.alterConfig(client, addOpts);
     }
 
-    @SuppressWarnings({"deprecation"})
-    static <T> Seq<T> seq(Collection<T> seq) {
-        return JavaConverters.asScalaIteratorConverter(seq.iterator()).asScala().toSeq();
+    private void verifyConfig(Admin client, Optional<String> brokerId, Map<String, String> config) throws Exception {
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:495]: Could we verify both name and value? for example:
```java
    private void verifyConfig(Admin client, Optional<String> brokerId, Map<String, String> config) throws Exception {
        ConfigResource configResource = new ConfigResource(ConfigResource.Type.BROKER, brokerId.orElse(defaultBrokerId));
        TestUtils.waitForCondition(() -> {
            Map<String, String> current = client.describeConfigs(singletonList(configResource)).all().get().values()
                .stream().flatMap(e -> e.entries().stream())
                .collect(Collectors.toMap(ConfigEntry::name, ConfigEntry::value));
            return config.entrySet().stream().allMatch(e -> e.getValue().equals(current.get(e.getKey())));
        }, 5000, config + " are not updated");
    }
```

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 13]
```diff
@@ -65,219 +74,479 @@
...
+                .forEach(key -> assertTrue(true, "Config key " + key + " is not found in the broker configs"));
+
+    }
+
+    private void deleteAndVerifyConfig(Admin client, Optional<String> brokerId, Set<String> config) throws Exception {
+        ConfigCommand.ConfigCommandOptions deleteOpts =
+                new ConfigCommand.ConfigCommandOptions(toArray(alterOpts, entityOp(brokerId),
+                        asList("--delete-config", String.join(",", config))));
+        ConfigCommand.alterConfig(client, deleteOpts);
+        verifyConfig(client, brokerId, Collections.emptyMap());
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: Could we add a new method to verify the "deleted" config has default value (the config `message.max.bytes` has default value)? for example:
```java
    private void verifyConfig(Admin client, Optional<String> brokerId, Set<String> config) throws Exception {
        ConfigResource configResource = new ConfigResource(ConfigResource.Type.BROKER, brokerId.orElse(defaultBrokerId));
        TestUtils.waitForCondition(() -> {
            Map<String, String> current = client.describeConfigs(singletonList(configResource)).all().get().values()
                .stream().flatMap(e -> e.entries().stream())
                .collect(Collectors.toMap(ConfigEntry::name, ConfigEntry::value));
            return config.stream().allMatch(current::containsKey);
        }, 5000, config + " are not updated");
    }
```

**Comment by m1a2st** [comment]
@chia7712, Please review, Thanks for your comments.

**Comment by chia7712** [comment]
@m1a2st Could you please fix conflicts?

**Comment by m1a2st** [comment]
@chia7712 , I have been resolve the conflict.

**Comment by chia7712** [comment]
@m1a2st please fix the build error

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 14]
```diff
@@ -65,219 +75,494 @@
...
+    private static String[] toArray(List<String>... lists) {
         return Stream.of(lists).flatMap(List::stream).toArray(String[]::new);
     }
 
-    public static String captureStandardErr(Runnable runnable) {
+    private String captureStandardMsg(Runnable runnable) {
         return captureStandardStream(true, runnable);
     }
 
-    private static String captureStandardStream(boolean isErr, Runnable runnable) {
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:263]: `isErr` is always true, right?

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 15]
```diff
@@ -65,219 +75,494 @@
...
+        Map<String, String> encoderConfigs = new HashMap<>(configs);
+        encoderConfigs.put(PASSWORD_ENCODER_SECRET_CONFIG, "encoder-secret");
+        alterConfigWithZk(zkClient, adminZkClient, Optional.of(brokerId), encoderConfigs);
         Properties brokerConfigs = zkClient.getEntityConfigs("brokers", brokerId);
-        assertFalse(brokerConfigs.contains(PasswordEncoderConfigs.PASSWORD_ENCODER_SECRET_CONFIG), "Encoder secret stored in ZooKeeper");
+        assertFalse(brokerConfigs.contains(PASSWORD_ENCODER_SECRET_CONFIG), "Encoder secret stored in ZooKeeper");
         assertEquals("2", brokerConfigs.getProperty("log.cleaner.threads")); // not encoded
-        String encodedPassword = brokerConfigs.getProperty("listener.name.external.ssl.keystore.password");
+        String encodedPassword = brokerConfigs.getProperty("listener.name.internal.ssl.keystore.password");
         PasswordEncoder passwordEncoder = ConfigCommand.createPasswordEncoder(JavaConverters.mapAsScalaMap(encoderConfigs));
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: in the production code, we convert the java collection to scala for `createPasswordEncoder`. By contrast, we do a reverse conversion in testing... that is weird. Could we change the input types from scala to java?

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 16]
```diff
@@ -65,219 +75,494 @@
...
         // Password config update without encoder secret should fail
         assertThrows(IllegalArgumentException.class,
-            () -> alterConfigWithZk(zkClient, Collections.singletonMap("listener.name.external.ssl.keystore.password", "secret"), Optional.of(brokerId)));
+                () -> alterConfigWithZk(zkClient, adminZkClient, Optional.of(brokerId),
+                        singletonMap("listener.name.internal.ssl.keystore.password", "secret")));
 
         // Password config update with encoder secret should succeed and encoded password must be stored in ZK
         Map<String, String> configs = new HashMap<>();
-        configs.put("listener.name.external.ssl.keystore.password", "secret");
+        configs.put("listener.name.internal.ssl.keystore.password", "secret");
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: why we need this change?

Comment (m1a2st) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: I read the Kafka document, I can't find the `listener.name.external.ssl.keystore.password`, but I can find the `listener.name.internal.ssl.keystore.password`, thus I considered that is wrong.

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java (hunk 14 above)
Comment (m1a2st) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:263]: Yes, I will remove this parameter

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java (hunk 16 above)
Comment (chia7712) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: `external` means there is a listener called `external` and we want to set its property `ssl.keystore.password`.

Comment (m1a2st) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: I was wrong. Appreciate your feedback.

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java (hunk 15 above)
Comment (m1a2st) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: Sorry, I don't understand why convert Java type to Scala is weird. If I change the input type in `ConfigCommand.scala` `createPasswordEncoder`, I should use `java.util.Map` in scala code, that isn't strange?

Comment (chia7712) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: sorry for unclear comment. My point was `createPasswordEncoder` should use java collection, since in production we do conversion from java to scala too. 
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/admin/ConfigCommand.scala#L248

Comment (m1a2st) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: I understand, Thanks for your comment.

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 17]
```diff
@@ -244,8 +248,11 @@ object ConfigCommand extends Logging {
           " to override the default encoding parameters. Password encoder configs will not be persisted" +
           " in ZooKeeper."
       )
-
-      val passwordEncoder = createPasswordEncoder(passwordEncoderConfigs.asScala)
+      val passwordConfigsMap = new java.util.HashMap[String, String]
```
Comment (m1a2st) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala:251]: I want use 
`passwordEncoderConfigs.entrySet()
  .stream()
  .collect(Collectors.toMap(_.getKey.toString, _.getValue.toString))`
  but Scala can't compile, so I use foreach to generate Java `Map<String, String>`

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 18]
```diff
@@ -210,15 +210,19 @@ object ConfigCommand extends Logging {
     }
   }
 
-  def createPasswordEncoder(encoderConfigs: Map[String, String]): PasswordEncoder = {
+  def createPasswordEncoder(encoderConfigs: java.util.Map[String, String]): PasswordEncoder = {
     encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_SECRET_CONFIG)
```
Comment (chia7712) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: It seems this line is unnecessary, right?

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 19]
```diff
@@ -210,15 +210,19 @@ object ConfigCommand extends Logging {
...
     PasswordEncoder.encrypting(new Password(encoderSecret),
       null,
-      encoderConfigs.getOrElse(PasswordEncoderConfigs.PASSWORD_ENCODER_CIPHER_ALGORITHM_CONFIG, PasswordEncoderConfigs.PASSWORD_ENCODER_CIPHER_ALGORITHM_DEFAULT),
-      encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_KEY_LENGTH_CONFIG).map(_.toInt).getOrElse(PasswordEncoderConfigs.PASSWORD_ENCODER_KEY_LENGTH_DEFAULT),
-      encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_ITERATIONS_CONFIG).map(_.toInt).getOrElse(PasswordEncoderConfigs.PASSWORD_ENCODER_ITERATIONS_DEFAULT))
+      encoderConfigs.getOrDefault(PasswordEncoderConfigs.PASSWORD_ENCODER_CIPHER_ALGORITHM_CONFIG, PasswordEncoderConfigs.PASSWORD_ENCODER_CIPHER_ALGORITHM_DEFAULT),
+      Optional.ofNullable(encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_KEY_LENGTH_CONFIG))
+        .map(Integer.parseInt)
+        .orElse(PasswordEncoderConfigs.PASSWORD_ENCODER_KEY_LENGTH_DEFAULT),
+      Optional.ofNullable(Integer.parseInt(encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_ITERATIONS_CONFIG)))
```
Comment (chia7712) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: If `PASSWORD_ENCODER_ITERATIONS_CONFIG` is undefined, `Integer.parseInt` will produce NPE, right?

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 20]
```diff
@@ -65,219 +74,488 @@
 @ExtendWith(value = ClusterTestExtensions.class)
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:73]: I guess `@SuppressWarnings("deprecation")` is unnecessary now.

File: core/src/main/scala/kafka/admin/ConfigCommand.scala (hunk 19 above)
Comment (m1a2st) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: Nope, if `encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_KEY_LENGTH_CONFIG)` is null, it will run `.orElse(PasswordEncoderConfigs.PASSWORD_ENCODER_KEY_LENGTH_DEFAULT)`, so NPE won't happen.
Under code NPE won't happened
```
Optional.ofNullable((String) null)
                .map(Integer::parseInt)
                .orElse(0);
```

Comment (chia7712) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: The return value from `encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_ITERATIONS_CONFIG)` is handled by `Integer.parseInt` first, so `Optional.ofNullable` can't wrap the NULL, right?

Comment (m1a2st) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: Sorry, I didn't see the last line. You are right!

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 21]
```diff
@@ -210,15 +210,19 @@ object ConfigCommand extends Logging {
...
+    val encoderSecret = Optional.ofNullable(encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_SECRET_CONFIG))
+      .orElseThrow(() => new IllegalArgumentException("Password encoder secret not specified"))
     PasswordEncoder.encrypting(new Password(encoderSecret),
       null,
-      encoderConfigs.getOrElse(PasswordEncoderConfigs.PASSWORD_ENCODER_CIPHER_ALGORITHM_CONFIG, PasswordEncoderConfigs.PASSWORD_ENCODER_CIPHER_ALGORITHM_DEFAULT),
-      encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_KEY_LENGTH_CONFIG).map(_.toInt).getOrElse(PasswordEncoderConfigs.PASSWORD_ENCODER_KEY_LENGTH_DEFAULT),
-      encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_ITERATIONS_CONFIG).map(_.toInt).getOrElse(PasswordEncoderConfigs.PASSWORD_ENCODER_ITERATIONS_DEFAULT))
+      encoderConfigs.getOrDefault(PasswordEncoderConfigs.PASSWORD_ENCODER_CIPHER_ALGORITHM_CONFIG, PasswordEncoderConfigs.PASSWORD_ENCODER_CIPHER_ALGORITHM_DEFAULT),
+      Optional.ofNullable(encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_KEY_LENGTH_CONFIG))
+        .map(Integer.parseInt)
```
Comment (chia7712) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: Please explicitly declare the type for scala 2.12. 

```scala
      Optional.ofNullable(encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_KEY_LENGTH_CONFIG))
        .map[Int](Integer.parseInt)
        .orElse(PasswordEncoderConfigs.PASSWORD_ENCODER_KEY_LENGTH_DEFAULT),
```

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 22]
```diff
@@ -210,15 +210,19 @@ object ConfigCommand extends Logging {
...
+        .orElse(PasswordEncoderConfigs.PASSWORD_ENCODER_KEY_LENGTH_DEFAULT),
+      Optional.ofNullable(encoderConfigs.get(PasswordEncoderConfigs.PASSWORD_ENCODER_ITERATIONS_CONFIG))
```
Comment (chia7712) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala:222]: ditto

**Review by chia7712 — APPROVED** [review]
LGTM
